src/solve/fulfill.rs

```diff
@@ -115,152 +133,235 @@ impl<'s> Fulfill<'s> {
-        debug!("fulfill::solve_one(wc={:?})", wc);
-
-        let quantified_wc = self.infer.make_query(&wc);
-        let solution = self.solver.solve(quantified_wc.clone())?;
-
-        // Regardless of whether the result is ambiguous or not,
-        // solving the where-clause may have yielded a refined
-        // goal. For example, if the original where-clause was
-        // something like `Foo<?4>: Borrow<?3>`, we would have
-        // "quantified" that to yield `exists ?0, ?1. Foo<?0>: Borrow<?1>`.
-        // We may now have gotten back a refined goal like `exists ?0. Foo<?0>:
-        // Borrow<Foo<?0>>`. In that case, we can unify `?3` with `Foo<?4>`.
-        //
-        // To make that unification happen, we first instantiate all
-        // the variables on the goal we got back with new inference
-        // variables. So we would thus convert `exists ?0. Foo<?0>:
-        // Borrow<Foo<?0>>` into `Foo<?5>: Borrow<Foo<?5>>`.  We would
-        // then unify this with our original goal (`Foo<?4>:
-        // Borrow<?3>`). This will result in the equations `?4 = ?5`
-        // and `?3 = Foo<?5>`.
-        //
-        // As a potential efficiency improvement, one could imagine a
-        // more algorithm written just for this case instead of
-        // instantiating with variables and applying the standard
-        // unification algorithm. But this is good enough for now.
-        let new_type_info = {
-            solution.refined_goal.binders != quantified_wc.binders ||
-            solution.refined_goal.value.value != quantified_wc.value
-        };
-
-        debug!("fulfill::solve_one: new_type_info={}", new_type_info);
-
-        if new_type_info || !solution.refined_goal.value.constraints.is_empty() {
-            let Constrained { constraints, value: refined_goal } =
-                self.instantiate(solution.refined_goal.binders.iter().cloned(),
-                                 &solution.refined_goal.value);
-
-            debug!("fulfill::solve_one: adding constraints {:?}", constraints);
-            self.constraints.extend(constraints);
-
-            debug!("fulfill::solve_one: unifying original and refined goal");
-            self.unify(&wc.environment, wc, &refined_goal)?;
-
-            if new_type_info {
-                *inference_progress = true;
+    fn fulfill_negative(&mut self) -> Result<Outcome> {
+        // Attempt to refute each negative obligations by in a fresh Fulfill
+        for goal in self.to_refute.drain(..) {
+            let canonicalized = self.infer.canonicalize(&goal);
+            let goal = canonicalized.quantified.value;
+
+            // Negation cannot be used to resolve existential variables, and do
+            // not have a useful (for us) logial meaning when they contain
+            // existential variables; treat negative goals with free variables
+            // as ambiguous
+            if !canonicalized.free_vars.is_empty() {
+                return Ok(Outcome::Incomplete);
```

Comment:
**Question:** This is stronger than necessary, right? It seems like any result of `Unique` with an empty substitution is probably good enough. Consider something like `not { Vec<?T>: Copy }`... we know that it is false (given closed-world, at least), simply because there are no impls of `Copy` *at all* that could apply.

(I made this comment earlier, but it got marked as "outdated" by GH for some reason.)